test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java

```diff
@@ -0,0 +1,210 @@
+                    .addToolArg("JVMTI.agent_load")
+                    .addToolArg(jvmtiAgentPath);
+            var pb = new ProcessBuilder(jcmd.getCommand());
+            int exitValue = ProcessTools.executeProcess(pb)
+                    .outputTo(System.out)
+                    .errorTo(System.out)
+                    .getExitValue();
+            assertEquals(0, exitValue);
+        };
+
+        // warning should be printed
+        test(op).shouldContain(message);
+
+        // no warning should be printed
+        test(op, "-XX:+EnableDynamicAgentLoading").shouldNotContain(message);
+    }
+
+    /**
+     * Test loading Java agent into a running VM.
+     */
+    @Test
+    void testLoadJavaAgent() throws Exception {
+        String message = "WARNING: A Java agent has been loaded dynamically";
+
+        // warning should be printed
+        test((pid, vm) -> vm.loadAgent(javaAgent)).shouldContain(message);
+
+        // no warning should be printed
+        test((pid, vm) -> vm.loadAgent(javaAgent), "-XX:+EnableDynamicAgentLoading")
+                .shouldNotContain(message);
+    }
+
+    private interface Op {
+        void accept(long pid, VirtualMachine vm) throws Exception;
+    }
+
+    /**
+     * Starts a new VM to run an application, attaches to the VM, runs a given action
+     * with the VirtualMachine object, and returns the output (both stdout and stderr)
+     * for analysis.
+     */
+    private OutputAnalyzer test(Op action, String... vmopts) throws Exception {
+        // start a listener socket that the application will connect to
+        try (ServerSocket listener = new ServerSocket()) {
+            InetAddress lh = InetAddress.getLoopbackAddress();
+            listener.bind(new InetSocketAddress(lh, 0));
+
+            var done = new AtomicBoolean();
+
+            // start a thread to wait for the application to phone home
+            Thread.ofPlatform().daemon().start(() -> {
+                try (Socket s = listener.accept();
+                     DataInputStream in = new DataInputStream(s.getInputStream())) {
+
+                    // read pid
+                    long pid = in.readLong();
+
+                    // attach and run the action with the vm object
+                    VirtualMachine vm = VirtualMachine.attach(""+pid);
+                    try {
+                        action.accept(pid, vm);
+                        done.set(true);
+                    } finally {
+                        vm.detach();
+                    }
+
+                    // shutdown
+                    s.getOutputStream().write(0);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            });
+
+            // launch application with the given VM options, waiting for it to terminate
+            Stream<String> s1 = Stream.of(vmopts);
+            Stream<String> s2 = Stream.of("Application", ""+listener.getLocalPort());
```

Comment:
Nit: Spaces are missed around '+' sign at lines: 121, 180 and 197.

Reply:
Okay, we can use Integer.toString and Long.toString instead.